lib/distinguished_name.ml

```diff
@@ -187,56 +191,58 @@ module Asn = struct
+      (X520.state_or_province_name   , fun x -> ST (of_c x)) ;
+      (X520.organization_name        , fun x -> O (of_c x)) ;
+      (X520.organizational_unit_name , fun x -> OU (of_c x)) ;
+      (X520.title                    , fun x -> T (of_c x)) ;
+      (X520.dn_qualifier             , fun x -> DNQ (of_c x)) ;
+      (PKCS9.email                   , fun x -> Mail (of_c x)) ;
+      (X520.given_name               , fun x -> Given_name (of_c x)) ;
+      (X520.surname                  , fun x -> Surname (of_c x)) ;
+      (X520.initials                 , fun x -> Initials (of_c x)) ;
+      (X520.pseudonym                , fun x -> Pseudonym (of_c x)) ;
+      (X520.generation_qualifier     , fun x -> Generation (of_c x)) ;
+      (X520.street_address           , fun x -> Street (of_c x)) ;
+      (userid                        , fun x -> Userid (of_c x))]
+      ~default:(fun oid x -> Other (oid, of_c x))
 
     and a_g = function
-      | DC x -> (domain_component, x )
-      | CN x -> (X520.common_name, x )
-      | Serialnumber x -> (X520.serial_number, x )
-      | C x -> (X520.country_name, x )
-      | L x -> (X520.locality_name, x )
-      | ST x -> (X520.state_or_province_name, x )
-      | O x -> (X520.organization_name, x )
-      | OU x -> (X520.organizational_unit_name, x )
-      | T x -> (X520.title, x )
-      | DNQ x -> (X520.dn_qualifier, x )
-      | Mail x -> (PKCS9.email, x )
-      | Given_name x -> (X520.given_name, x )
-      | Surname x -> (X520.surname, x )
-      | Initials x -> (X520.initials, x )
-      | Pseudonym x -> (X520.pseudonym, x )
-      | Generation x -> (X520.generation_qualifier, x )
-      | Street x -> (X520.street_address, x )
-      | Userid x -> (userid, x )
-      | Other (oid, x) -> (oid, x )
+      | DC x -> (domain_component, `C3 x )
+      | CN x -> (X520.common_name, `C1 x )
+      | Serialnumber x -> (X520.serial_number, `C2 x )
+      | C x -> (X520.country_name, `C2 x )
+      | L x -> (X520.locality_name, `C1 x )
+      | ST x -> (X520.state_or_province_name, `C1 x )
+      | O x -> (X520.organization_name, `C1 x )
+      | OU x -> (X520.organizational_unit_name, `C1 x )
+      | T x -> (X520.title, `C1 x )
+      | DNQ x -> (X520.dn_qualifier, `C2 x )
+      | Mail x -> (PKCS9.email, `C3 x )
+      | Given_name x -> (X520.given_name, `C1 x )
+      | Surname x -> (X520.surname, `C1 x )
+      | Initials x -> (X520.initials, `C1 x )
+      | Pseudonym x -> (X520.pseudonym, `C1 x )
+      | Generation x -> (X520.generation_qualifier, `C1 x )
+      | Street x -> (X520.street_address, `C1 x )
+      | Userid x -> (userid, `C1 x )
+      | Other (oid, x) -> (oid, `C1 x )
     in
 
     let attribute_tv =
       map a_f a_g @@
       sequence2
         (required ~label:"attr type"  oid)
         (* This is ANY according to rfc5280. *)
-        (required ~label:"attr value" directory_name)
+        (required ~label:"attr value" directory_name2)
```

Comment:
@hannesm I can make a PR to mainline, but i'm not sure how it should be done the right way - attribute value should be different type according to oid and we have here the same problem with asn1 grammar ambigousity - country name should be `printable_name`, while organization name - `directory_name` that already contains `printable_name` in its choice

Reply:
if you PR this to mirleft/ocaml-x509 I'll try to find some time to review. this PR is a pretty neat solution (backwards compatible, and adheres to the standard) for https://github.com/mirleft/ocaml-x509/issues/69 -- but as noted in https://github.com/mirleft/ocaml-x509/issues/69#issuecomment-747316426 we may need a path to preserve the string tag of the subject / issuer (your `of_c` throws this away) --> this won't be fully backwards-compatible.